pkg/sidecar/server.go

```diff
@@ -79,10 +80,15 @@ func exportMetrics(metrics *dnsmasq.Metrics) {
 		// previous to get the proper value. This is needed because the
 		// Counter API does not allow us to set the counter to a value.
 		previousValue := countersCache[key]
-		delta := float64((*metrics)[key]) - previousValue
-		newValue := previousValue + delta
-		// Update cache to new value.
-		countersCache[key] = newValue
-		counters[key].Add(delta)
+		newValue := float64((*metrics)[key])
+		countersCache[key] = math.Max(newValue, 0)
```

Comment:
If the previous value were 200 and new value is 250, would this set the value to 250 and in line 91, make the value 250 + (250 - 200) = 300?
That's not what we want, correct? 
Also, could you run this and trigger a dnsmasq restart and make sure sidecar exports the correct metrics?

Reply:
No I don't think so, I believe this is functionally identical. Previously with the two variables `delta` and `newValue` there was an odd bit of math:

```go
delta := float64((*metrics)[key]) - previousValue
newValue := previousValue + delta
```
Which is equal to:
```go
newValue := previousValue + (float64((*metrics)[key]) - previousValue)
```
The `previousValue` variables cancel out. So it is just a redundant bit of math I got rid of. 

Also regarding _running this and triggering a restart_, that is well outside of my experience as a Kubernetes (and linux dnsmasq) dev. I would need some guidance on how to do that. Something like:

1. rebuild the container
2. run it with docker or something?
3. ssh into the container and exec some magic command that forces a dnsmasq restart?

Reply:
Yes those are the steps. 
1) rebuild container and push to a dev registry with `make push REGISTRY=<name>`
2) I usually run this on an existing kubernetes cluster, i just change the image to the one i built. then ssh into dnsmasq and send SIGTERM via kill. 

Regarding the counter values, that explanation makes sense. So in case of reset, we will continue to increment dnsmasq stats from its previous value from before the crash. Would be good to verify it actually happens.

Reply:
Okay I will do those steps and attempt to test it next week on probably Tuesday. Thanks and hang tight.

Reply:
Hey @prameshj, sorry for being a month late with this. I have done the above steps and repeated the steps which previously caused crash looping in the DNS sidecar. There are now 0 pod restarts and the panics are gone from the logs. I think this is safe to submit. :)

Reply:
Thanks for checking. With this change, metrics will be reported when the new value is greater than the old value. If a dnsmasq restart happens and the value resets, no new metrics will be reported, until they catch up to the older value before dnsmasq restarted. Correct?

Reply:
No, previous value is always updated to `math.Max(newValue, 0)` no matter if it is greater than the previous value.

So if for instance the state is:

```go
countersCache[key] == 5 // previousValue will be set to 5 in line 82
```

If for the next two values for this key are `1` and `2` then the following will happen. For the call when `(*metrics)[key]` is equal to `1`:

```go
previousValue := countersCache[key] // 5
newValue := float64((*metrics)[key]) // 1
countersCache[key] = math.Max(newValue, 0) // updates countersCache[key] to 1
if newValue > previousValue { // 1 is not greater than 5 so we don't enter
  // counter add
}
```

Next for the call when `(*metrics)[key]` is equal to `2`:

```go
previousValue := countersCache[key] // 1
newValue := float64((*metrics)[key]) // 2
countersCache[key] = math.Max(newValue, 0) // updates countersCache[key] to 2
if newValue > previousValue { // 2 is greater than 1 so we enter
  // Counter .Add(...) with delta of (2-1)
}
```

So for that first instance where we have a huge decrease in value we don't log the negative delta, but in all subsequent calls we will take that new lower value as the next baseline.

Reply:
yes, makes sense. I missed that you update previous value every time.